packet_encoder.go

```diff
@@ -22,6 +22,13 @@ type packetEncoder interface {
 	// Stacks, see PushEncoder
 	push(in pushEncoder)
 	pop() error
+
+	// Statistics in order to expose metrics
+	offset() int
+	recordTopic(topic string)
```

Comment:
I agree this add unwanted coupling but I am sure not sure how to be aware of a context (topic here) when recording statistics.
What cleaner approach would you suggest?